tests/scanner/audit/retention_rules_engine_test.py

```diff
@@ -0,0 +1,856 @@
+        expect_violations = frsd.build_bucket_violations(
+            fake_bucket, 'only min retention')
+        self.assertEqual(got_violations, expect_violations)
+
+    yaml_str_both_min_and_max_retention = """
+rules:
+  - name: both min and max retention
+    applies_to:
+      - bucket
+    resource:
+      - type: bucket
+        resource_ids:
+          - fake_bucket
+    minimum_retention: 90
+    maximum_retention: 365
+
+"""
+
+    def test_both_min_max_no_condition(self):
+        """Test that a rule with both minimum_retention and maximum_retention
+        produces violations. A bucket's rule break it, if the bucket breakes the
+        maximum_retention part."""
+        rules_engine = get_rules_engine_with_rule(RetentionRulesEngineTest.yaml_str_both_min_and_max_retention)
+        self.assertEqual(1, len(rules_engine.rule_book.resource_rules_map))
+
+        data_creater = frsd.FakeBucketDataCreater('fake_bucket', frsd.PROJECT1)
+
+        fake_bucket = data_creater.get_resource()
+        got_violations = list(rules_engine.find_violations(fake_bucket))
+        expect_violations = frsd.build_bucket_violations(
+            fake_bucket, 'both min and max retention')
+        self.assertEqual(got_violations, expect_violations)
+
+    def test_both_min_max_normal_del_any_del(self):
+        """Test that a rule with both minimum_retention and maximum_retention
+        produces violations. A bucket's rule break it, if the bucket breakes the
+        minimum_retention part."""
+        rules_engine = get_rules_engine_with_rule(RetentionRulesEngineTest.yaml_str_both_min_and_max_retention)
+        self.assertEqual(1, len(rules_engine.rule_book.resource_rules_map))
+
+        data_creater = frsd.FakeBucketDataCreater('fake_bucket', frsd.PROJECT1)
+        data_creater.AddLifecycleDict(action="Delete", age=100)
+        data_creater.AddLifecycleDict(action="Delete", is_live=True)
+
+        fake_bucket = data_creater.get_resource()
+        got_violations = list(rules_engine.find_violations(fake_bucket))
+        expect_violations = frsd.build_bucket_violations(
+            fake_bucket, 'both min and max retention')
+        self.assertEqual(got_violations, expect_violations)
+
+    def test_both_min_max_normal_del(self):
+        """Test that a rule with both minimum_retention and maximum_retention
+        does not produce violations."""
+        rules_engine = get_rules_engine_with_rule(RetentionRulesEngineTest.yaml_str_both_min_and_max_retention)
+        self.assertEqual(1, len(rules_engine.rule_book.resource_rules_map))
+
+        data_creater = frsd.FakeBucketDataCreater('fake_bucket', frsd.PROJECT1)
+        data_creater.AddLifecycleDict(action="Delete", age=100)
+
+        fake_bucket = data_creater.get_resource()
+        got_violations = list(rules_engine.find_violations(fake_bucket))
+        self.assertEqual(got_violations, [])
+
+    def test_both_min_max_3_conditions(self):
```

Comment:
what do the numbers in the test names mean? can we use descriptive names instead of numbers?

Reply:
It means there are 3 conditions.